### PR TITLE
Change oauth2 plugin auth. challenge to follow rfc specs

### LINKED ISSUE #737: Invalid OAuth2 challenge-response authentication implementation
Kong OAuth2 plugin doesn't follow standards specified in 
* [rfc2616 – HTTP/1.1](http://tools.ietf.org/html/rfc2616)
* [rfc2617 – HTTP Authentication](http://tools.ietf.org/html/rfc2617) and 
* [rfc6750 – The OAuth 2.0 Authorization Framework: Bearer Token Usage](https://tools.ietf.org/html/rfc6750). 

The current OAuth2 plugin responses are forcing clients to implement hacks, and/or incorrect implementations for handling authentication challenges. This is especially problematic when using a generic client implementations that handle OAuth2 authentication challenges for services behind kong gateway.

## Response to a request if it lacks any authentication information

Currently kong response is `403 Forbidden`

> The server understood the request, but is refusing to fulfill it.
   Authorization will not help and the request SHOULD NOT be repeated.
  [rfc2616#section-10.4.4](http://tools.ietf.org/html/rfc2616#section-10.4.4) 

This response is misleading. It doesn't even contradict the 403 by having any other why of telling the client that the endpoint requires user authentication e.g. `WWW-Authenticate`. If the client was unaware that authentication is necessary it may interpret the response as an application error and think that the request was otherwise legal.

**Correct response** would be `401 Unauthorized` with `WWW-Authenticate: Bearer realm="example"` header 
> The request requires user authentication. The response MUST include a
   WWW-Authenticate header field ... The client MAY repeat the
   request with a suitable Authorization header field
   [rfc2616#section-10.4.2](http://tools.ietf.org/html/rfc2616#section-10.4.2)

## Response to a request with Authorization header containing the access token that is expired, revoked, malformed, or invalid for other reasons

Kong's response to expired token the is `400 Bad Request`: 
> The request could not be understood by the server due to malformed
   syntax. The client SHOULD NOT repeat the request without
   modifications.
   [rfc2616#section-10.4.1](http://tools.ietf.org/html/rfc2616#section-10.4.1)

and in case of an invalid token it is `403 Forbidden`. Both responses do not follow the standards. 

[rfc2617#section-1.2](http://tools.ietf.org/html/rfc2617#section-1.2) gives a very clear description on how the server should respond in this case.

> If the origin server does not wish to accept the credentials sent
   with a request, it SHOULD return a 401 (Unauthorized) response. The
   response MUST include a WWW-Authenticate header field containing at
   least one (possibly new) challenge applicable to the requested
   resource.

In addition [rfc6750#section-3.1](https://tools.ietf.org/html/rfc6750#section-3.1) clearifies the error codes and specifies appropriate HTTP status codes.

**Correct response** would again be in both cases `401 Unauthorized` with `WWW-Authenticate: Bearer realm="example", error="invalid_token", error_description="The access token expired/The access token invalid/etc"`

## PR

I'm willing work on a PR to fix the issues mentioned above to make the OAuth2 plugin follow the standards. I'm aware that changing the implementation without maintaining backward compatible could be an issue, but trying to maintain BC can be very painful and unnecessarily bloat the code. 

Please let me know what you think and then lets take it from there.

### PR DESCRIPTION
**Pull request by CyExy**
closes #737

### CODE REVIEW

File: kong/plugins/oauth2/access.lua [hunk 1]
```diff
@@ -376,16 +376,21 @@ function _M.execute(conf)
...
-  local token = retrieve_token(parse_access_token(conf))
+  local accessToken = parse_access_token(conf);
+  if not accessToken then
+    return responses.send_HTTP_UNAUTHORIZED({}, false, {["WWW-Authenticate"] = 'Bearer realm="service"'})
+  end
+
+  local token = retrieve_token(accessToken)
   if not token then
-    return responses.send_HTTP_FORBIDDEN("Invalid authentication credentials")
+    return responses.send_HTTP_UNAUTHORIZED({[ERROR] = "invalid_token", error_description = "The access token invalid"}, false, {["WWW-Authenticate"] = 'Bearer realm="service" error="invalid_token" error_description="The access token invalid"'})
```
Comment (thefosk) [inline kong/plugins/oauth2/access.lua]: I see a typo, should be: "The access token **is** invalid".

Comment (thefosk) [inline kong/plugins/oauth2/access.lua]: On a side note, I don't see `invalid_token` as one of the possible error types, as specified in the spec: https://tools.ietf.org/html/rfc6749#section-5.2

Comment (CyExy) [inline kong/plugins/oauth2/access.lua]: Thanks for the typo, I'll do a rebase.

**Comment by CyExy** [comment]
> On a side note, I don't see `invalid_token` as one of the possible error types, as specified in the spec: https://tools.ietf.org/html/rfc6749#section-5.2

You are looking at the errors for authorization server, as [rfc6749#section-1.4](https://tools.ietf.org/html/rfc6749#section-1.4) says `Access token attributes and the
   methods used to access protected resources are beyond the scope of
   this specification and are defined by companion specifications such
   as [RFC6750].` 

You can find them here: https://tools.ietf.org/html/rfc6750#section-3.1

**Comment by thefosk** [comment]
@CyExy nice, thanks for the clarification. This PR looks good to me.

**Pull request merged**